### DISCUSSION TITLE
Question about time varying profile of the species

### DISCUSSION THREAD

**Discussion by leo256cb** [epochpic/epoch#637] [Q&A]
Hi,

I'm a new user of epoch

May I ask if it's possible using EPOCH to creates a specie which have time varying number density characteristics? I know that the injector can inject particle with a time profile but I think the injector only allows to inject particles from the domain boundaries. 

In the current 2D simulation I'm doing, instead of injecting from the boundaries I am trying to directly introduce a package of electrons in the middle of the domain with a Gaussian number density profile in both time and space. From my visualization results I think the spatial Gaussian implementation is successful but I couldn't get the time profile to work in the spices block. May I ask if EPOCH will actually allow time varying number density characteristics in species block?

I think I have found a old discussion similar to the issue I have but the discussion don't have an answer yet, link:
[https://github.com/Warwick-Plasma/epoch/discussions/128?sort=old](url)

Thank you for reading my question and I have attached the species block I set up.

```
# Particles added here, instead of using injectors,
# I would like to introduce a electrons package directly in the middle of the domain with momentum as drift_p
begin:species
  name = Electron_package
  charge = -1
  mass = 1.0
  frac = 0.5
  dump = T
  
  # the exp(-(x^2+y^2)/(5e-8)^2) gives the spatial gussian profile and can I use gauss(time, 50e-15, 50e-15) to give the specie's density a time profile?
  number_density = if (((x^2)+(y^2)) gt ((10e-8)^2), 0, 1.6013e19*exp(-(x^2+y^2)/(5e-8)^2)*gauss(time, 50e-15, 50e-15))
  
  # drift_p set in the constant block
  drift_x = drift_p
end:species
```

Sincerely,
Leo

**Comment by leo256cb** [comment]
Ohh sorry I found the link I attached above is not working, I will reattached it here:
Link for old discussion of similar issue: **https://github.com/Warwick-Plasma/epoch/discussions/128**

Sincerely,
Leo

**Comment by Status-Mirror** [comment]
Hey Leo,

The number density is only used for the initial particle load I'm afraid. While it might accept a `time` value in the function, it will only use the time at the start of the simulation for the initial load. There's no EPOCH feature I'm aware of which can provide the functionality you seek.

If you feel confident coding in FORTRAN, there may be a way to tweak the source-code to give you your desired behaviour. I think the easiest approach here would be to modify the `src/physics_packages/file_injector.F90 file`. Currently in EPOCH2D, it adds particles to the simulation using the weight, time, momentum and position on a given boundary. You could change the script to force it to accept both $x$ and $y$ injection positions simultaneously, then modify the `run_file_injection` subroutine to assign these variables to the loaded particle position. For an introduction to file injectors, check out our [file injector demo](https://epochpic.github.io/quickstart/basic_examples/injectors.html#file-injectors).

You'll need to add an additional check to ensure the particle has been injected into the correct MPI rank, but there are lines of code doing similar things already in that subroutine.

Hope this helps,
Stuart

**Reply by leo256cb** [reply]
Hello Stuart,

Thank you very much for your replies and clarification, I will try to look into the source code.

Sincerely,
Leo

**Reply by leo256cb** [reply]
Hello Stuart (@Status-Mirror),

Thank you for your previous reply. I changed certain parts in the `src/physics_packages/file_injector.F90` file but when I ran the simulation for some reason the injector still only inject particles from the x_min boundary (which is the boundary I set in the input.deck). 

Since the `file_injector.F90` code will read .txt files contain the information of particle positions (`injection_x.txt` and `injection_y.txt`), weight (`injection_w.txt`), momentum (`injection_px.txt`) etc, I created all those .txt files using MATLAB and I believe the EPOCH opened and read-in those files successfully based on the feedback printed on my terminal. But for some reason the data from `injection_x.txt` (for specify the x-position of the particles) is ignored afterwards and the injector still inject the particle on the x_min boundary.

May I ask do you mind to give it a look on the changes I have made for the `file_injector.F90` FORTRAN code? I put comments in the code for my changes with a indicator "***Modified***" (if you ctrl+F to search the word "***Modified***", then you can see the location on where I made the changes). I must be missing something in the code which will automatically set all the x-position to x_min when it sees my boundary setup in the injector.

I attached both the edited and original version of the `file_injector.F90` below in a .zip folder. I named them `file_injector_edited.F90` and `file_injector_original.F90` respectively.
[file_injectors_code.zip](https://github.com/Warwick-Plasma/epoch/files/14517719/file_injectors_code.zip)

Sorry for the inconvenience but this is the first time I try to modify a FORTRAN code. 

Thank you for reading my reply.

Sincerely,
Leo
